### PR TITLE
posix: eventfd: fix waking up poll()

### LINKED ISSUE #59577: posix: eventfd: poll(..., POLLIN, ...) does not wake up after `eventfd_write()`
**Describe the bug**
When on thread is waiting for event from `eventfd` mechanism using `poll()` and the other thread generates an event with `eventfd_write()`, then the former is not woken up (i.e. `poll()` does not return).

This is a regression introduced by e6eb0a705bc8d92411d134bf3e7c2f82df336231, which removed `read_sig` and `write_sig` (mechanisms used for waking up `poll()`), stating that those were unnecessary.

**To Reproduce**
1. Create one eventfd: `fd = eventfd(0, EFD_NONBLOCK);`
2. Create `fds`: `struct pollfd fds[] = { .fd = fd, .events = POLLIN };`
3. Thread 1: call `poll(fds, 1, -1);`
4. Thread 2: call `eventfd_write(fd, 1);`
5. Thread 1 never wakes up

**Expected behavior**
`poll()` should return after another thread calls `eventfd_write()`.

**Impact**
Showstopper for applications relying on `poll()` behavior for `eventfd` file descriptors. Since this is a regression in 3.4.0, it stops such applications from using Zephyr 3.4.0 and newer.

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows)
 - Toolchain: Zephyr SDK 0.16.0
 - Commit SHA or Version used: v3.4.0 and tip of main (c6c93e394d6a52c4541d39135681fc81b8df7dd5)

### PR DESCRIPTION
**Pull request by mniestroj**
Fix a regression introduced by commit e6eb0a705bc8 ("posix: eventfd: revise
locking, signaling, and allocation"), which was a complete rewrite stating
that:

> The `wait_q` and `k_poll_signal` entries were removed from
> `struct eventfd` as they were unnecessary.

In fact, `k_poll_signal` (both `read_sig` and `write_sig`) were used to
wake-up blocking `poll()` invocation in another thread. This is no longer
the case now, i.e. `poll(..., POLLIN)` does not return after calling
`eventfd_write()` on the observed (polled) FD.

Fix this regression by bringing back `read_sig` and `write_sig` to very
similar state as it was before.

Fixes: e6eb0a705bc8 ("posix: eventfd: revise locking, signaling, and allocation")
Fixes: #59577

### CODE REVIEW

File: lib/posix/eventfd.c [hunk 1]
```diff
@@ -17,6 +17,8 @@
 #define EFD_FLAGS_SET_INTERNAL (EFD_SEMAPHORE | EFD_NONBLOCK)
 
 struct eventfd {
+	struct k_poll_signal read_sig;
```
Comment (cfriedt) [inline lib/posix/eventfd.c:20]: Hmmm.. I was really hoping to avoid these two. Is there no other `poll()` notification mechanism?

**Comment by cfriedt** [comment]
@mniestroj - you will need to create an issue for your backport

**Comment by cfriedt** [comment]
Overall, not a significant performance degradation from what I can tell, so the complete rewrite and speedup of ~10x was probably worth it 😉

Thank you for your contribution to the testsuite.

**Review by cfriedt — APPROVED** [review]

File: lib/posix/eventfd.c (hunk 1 above)
Comment (mniestroj) [reply lib/posix/eventfd.c:20]: AFAIK this is one of the most lightweight mechanisms. At least nothing better comes to my mind right now.

**Review by jgl-meta — APPROVED** [review]

**Pull request merged**